File: GCDTP/frontend/src/api/rootCause.js
/**
 * Root Cause Analysis API
 * 
 * API client for Root Cause Analysis Engine.
 * READ ONLY - NO automation, NO work orders.
 */

const API_BASE_URL = process.env.REACT_APP_API_URL || 'http://localhost:8080';

// Analyze asset
export async function analyzeAsset(assetId, params = {}) {
  const searchParams = new URLSearchParams({
    analysis_type: params.analysis_type || 'failure',
    event_id: params.event_id || '',
    time_window_hours: params.time_window_hours || 24
  });
  
  // Remove empty event_id
  if (!params.event_id) {
    searchParams.delete('event_id');
  }

  const response = await fetch(
    `${API_BASE_URL}/root-cause/analyze/${assetId}?${searchParams}`,
    { method: 'POST' }
  );
  
  if (!response.ok) throw new Error('Failed to analyze asset');
  return response.json();
}

// Get analysis by ID
export async function getAnalysis(analysisId) {
  const response = await fetch(`${API_BASE_URL}/root-cause/${analysisId}`);
  if (!response.ok) throw new Error('Failed to get analysis');
  return response.json();
}

// Get analyses for asset
export async function getAnalysesForAsset(assetId, limit = 10) {
  const response = await fetch(
    `${API_BASE_URL}/root-cause/asset/${assetId}?limit=${limit}`
  );
  if (!response.ok) throw new Error('Failed to get analyses');
  return response.json();
}

// Analyze event
export async function analyzeEvent(eventId, assetId) {
  const response = await fetch(
    `${API_BASE_URL}/root-cause/event/${eventId}?asset_id=${assetId}`,
    { method: 'GET' }
  );
  if (!response.ok) throw new Error('Failed to analyze event');
  return response.json();
}

// Get factors for analysis
export async function getFactors(analysisId) {
  const response = await fetch(`${API_BASE_URL}/root-cause/factors/${analysisId}`);
  if (!response.ok) throw new Error('Failed to get factors');
  return response.json();
}

// Get chains for analysis
export async function getChains(analysisId) {
  const response = await fetch(`${API_BASE_URL}/root-cause/chains/${analysisId}`);
  if (!response.ok) throw new Error('Failed to get chains');
  return response.json();
}

// Get high confidence analyses
export async function getHighConfidenceAnalyses(threshold = 0.7) {
  const response = await fetch(
    `${API_BASE_URL}/root-cause/high-confidence?threshold=${threshold}`
  );
  if (!response.ok) throw new Error('Failed to get high confidence analyses');
  return response.json();
}

// Get analysis history
export async function getAnalysisHistory(limit = 20) {
  const response = await fetch(
    `${API_BASE_URL}/root-cause/history?limit=${limit}`
  );
  if (!response.ok) throw new Error('Failed to get history');
  return response.json();
}

export default {
  analyzeAsset,
  getAnalysis,
  getAnalysesForAsset,
  analyzeEvent,
  getFactors,
  getChains,
  getHighConfidenceAnalyses,
  getAnalysisHistory
};

File: GCDTP/backend/src/schemas/root_cause.py
"""
Root Cause Analysis Schemas

Pydantic schemas for Root Cause Analysis Engine.
READ ONLY - NO automation, NO work orders.
"""

from typing import Optional, List, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field


# Cause Factor Schema
class CauseFactorSchema(BaseModel):
    """Schema for cause factor."""
    id: str
    analysis_id: str
    factor_type: str
    reference_id: Optional[str] = None
    weight: float
    description: str
    evidence: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


# Cause Chain Schema
class CauseChainSchema(BaseModel):
    """Schema for cause chain."""
    id: str
    analysis_id: str
    depth: int
    source_asset_id: str
    target_asset_id: str
    relationship_type: Optional[str] = None
    description: Optional[str] = None
    propagation_time_seconds: Optional[int] = None
    created_at: datetime

    class Config:
        from_attributes = True


# Root Cause Response Schema
class RootCauseResponse(BaseModel):
    """Schema for root cause response."""
    id: str
    asset_id: str
    event_id: Optional[str] = None
    analysis_type: str
    probable_cause: str
    confidence: float
    summary: Optional[str] = None
    impacted_assets: List[str] = []
    root_assets: List[str] = []
    created_at: datetime
    factors: List[CauseFactorSchema] = []
    chains: List[CauseChainSchema] = []

    class Config:
        from_attributes = True


# Factor Response Schema
class FactorResponse(BaseModel):
    """Schema for factor response."""
    id: str
    factor_type: str
    description: str
    weight: float
    evidence: Optional[str] = None


# Ranked Factor Response Schema
class RankedFactorResponse(BaseModel):
    """Schema for ranked factor."""
    rank: int
    factor: FactorResponse


# Cause Chain Response Schema
class CauseChainResponse(BaseModel):
    """Schema for cause chain response."""
    id: str
    depth: int
    source_asset_id: str
    target_asset_id: str
    relationship_type: Optional[str] = None
    description: Optional[str] = None
    propagation_time_seconds: Optional[int] = None


# Confidence Response Schema
class ConfidenceResponse(BaseModel):
    """Schema for confidence analysis."""
    overall_confidence: float
    factor_score: float
    chain_score: float
    evidence_score: float
    factor_count: int
    chain_depth: int
    evidence_count: int


# Asset Cause Summary Schema
class AssetCauseSummary(BaseModel):
    """Schema for asset cause summary."""
    asset_id: str
    asset_name: str
    analysis_count: int
    avg_confidence: float
    max_confidence: float
    common_causes: List[str] = []
    last_analysis: Optional[datetime] = None


# High Confidence Analysis Schema
class HighConfidenceAnalysis(BaseModel):
    """Schema for high confidence analysis."""
    id: str
    asset_id: str
    asset_name: Optional[str] = None
    analysis_type: str
    probable_cause: str
    confidence: float
    summary: Optional[str] = None
    created_at: datetime


# Analysis Request Schema
class AnalysisRequest(BaseModel):
    """Schema for analysis request."""
    asset_id: str
    analysis_type: str = "failure"
    event_id: Optional[str] = None
    time_window_hours: int = Field(default=24, ge=1, le=720)


# Analysis Response Schema
class AnalysisResponse(BaseModel):
    """Schema for analysis response."""
    analysis: RootCauseResponse
    confidence: ConfidenceResponse
    ranked_factors: List[RankedFactorResponse] = []
    chain: List[CauseChainResponse] = []


# Event Sequence Schema
class EventSequenceItem(BaseModel):
    """Schema for event in sequence."""
    timestamp: datetime
    event_type: str
    asset_id: str
    severity: str
    description: str


# Event Sequence Response Schema
class EventSequenceResponse(BaseModel):
    """Schema for event sequence response."""
    asset_id: str
    events: List[EventSequenceItem]
    total_events: int


# Dependency Trace Schema
class DependencyTraceItem(BaseModel):
    """Schema for dependency trace item."""
    depth: int
    asset_id: str
    asset_name: str
    relationship_type: str
    health_score: Optional[float] = None


# Dependency Trace Response Schema
class DependencyTraceResponse(BaseModel):
    """Schema for dependency trace response."""
    root_asset_id: str
    trace: List[DependencyTraceItem]
    total_depth: int
